### DISCUSSION TITLE
Deserialize a property to its subclass correctly

### DISCUSSION THREAD

**Discussion by wjchicago** [pydantic/pydantic#2749] [Question] [answered]
Dear all,

I checked all open issues and seems cannot find answer for this. Please direct me to the right thread if this is a duplicate.

**Problem description**
When a property of a class is another pydantic model's subclass, the serialization-deserialization round trip will not return to its original value.

**Code snipet**

```
import json
from pydantic import BaseModel

class FooBase(BaseModel):
    a: str
    b: int

class Foo(FooBase):
    c: bool

class Bar(BaseModel):
    d: FooBase

x = Foo(a='a', b=1, c=True)
y = Bar(d=x)

assert y == Bar.parse_raw(y.json())
```
The assertion will fail because property d is truncated to FooBase

The closest solution i can find for this is to bypass the validation by calling the construct() method, say

`assert y == Bar.construct(**json.loads(y.json()))`

Is there a better way to deal with this?

**Comment by PrettyWood** [comment, marked as answer]
Hey @wjchicago 

If you really want to get back the instance, you should probably do
```py
class Foo1(FooBase): ...
class Foo2(FooBase): ...

class Base(BaseModel):
    d: Union[Foo1, Foo2]
```
If you really want to set `d: FooBase` but still get back the subclass, you need to register them and find a way to get them back.
You can find an example [here](https://github.com/samuelcolvin/pydantic/issues/2177#issuecomment-739578307)

Hope it helps!

**Reply by wjchicago** [reply]
Appreciate it!
